lib/error/try_assign.h

```diff
@@ -0,0 +1,88 @@
+#ifndef LIB_ERROR_TRY_ASSIGN_H_
+#define LIB_ERROR_TRY_ASSIGN_H_
+
+#include "lib/base/paste.h"
+#include "lib/error/try.h"
+
+#define _TRY_ASSIGN(lvalue, expr, tmp) \
+  auto tmp = (expr);                   \
```

Comment:
so:
```
    int a = 5;
    if (Ready())
      TRY_ASSIGN(a, OperationStatus)
```
breaks. You need some sort of block statement? How can that work when introducing a new type? for lvalue = ...? Only trick that comes to mind is a gcc extension, eg, something like:

```
lvalue = ({ TRY(GetStatus(tmp)); GetValue(std::move(tmp)); })
```

Same comment as above about const ref. I'd also prefer the tmp variable to be declare in its own scope.

Reply:
Great point about if() TRY_ASSIGN();. I consider any 'if' without {} to be an error so I wasn't even considering that.

I had no idea about this GCC extension. It is exactly what this needs to be more robust. It can also remove the need for COUNTER. I never looked at the implementation of ASSIGN_OR_RETURN, so I'm not sure how it was handled there.